core/state_transition.go

```diff
@@ -66,10 +66,14 @@ func (result *ExecutionResult) Revert() []byte {
 	return common.CopyBytes(result.ReturnData)
 }
 
-// IntrinsicGas computes the 'intrinsic gas' for a message with the given data.
-func IntrinsicGas(data []byte, accessList types.AccessList, authList []types.SetCodeAuthorization, isContractCreation, isHomestead, isEIP2028, isEIP3860 bool) (uint64, error) {
+// IntrinsicGas computes the 'intrinsic gas' and the number of tokens for EIP-7623
+// for a message with the given data.
+func IntrinsicGas(data []byte, accessList types.AccessList, authList []types.SetCodeAuthorization, isContractCreation, isHomestead, isEIP2028, isEIP3860 bool) (uint64, uint64, error) {
```

Comment:
I think this is overloading `IntrinsicGas(..)`. The intrinsic gas of the tx is the amount paid upfront before execution starts. A large portion of this is the number of zero / non-zero bytes, however, tokens just end up as an unrelated byproduct of the calculation.

Iterating the data byte-by-byte twice is annoying, but I think if that is a perf bottleneck we could just calculate the total number of zero bytes ahead of time and feed that to a new intrinsic gas / data floor gas function that operate on zero bytes and data len instead of entire data field.

Reply:
So I was a bit confused as here you are talking about calculating the total number of zero bytes ahead of time and passing that in. Whereas later you suggested a different approach.

However I had a look and figured out I can instead of tokens return the floor gas itself from `IntrinsicGas`. Given that it is a sort of second type of "intrinsic gas" it seems satisfactory to me. It also made the code generally cleaner (as gas floor computed once).

Please let me know what you think.

Reply:
I kinda prefer the way @s1na wrote it as well. It forces the user of IntrinsicGas to think about it, while the diff that you posted will not. This way users will make mistakes (like forgetting to do the calculation in the txpool or in t8n) as they need to have the knowledge that after the IntrinsicGas calculation, the DataGas calculation has to follow